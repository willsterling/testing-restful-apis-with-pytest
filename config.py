BASE_URI = 'https://todo.pixegami.io'
BOOKER_URI = 'https://restful-booker.herokuapp.com'