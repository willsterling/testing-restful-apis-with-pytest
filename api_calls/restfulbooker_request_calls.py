import requests
from config import BOOKER_URI

# HELPER FUNCTIONS
# create token

# create authorization (credentials below are required for successful creation of token)
def new_auth():
    return {
        "username": "admin",
        "password": "password123"
    }

def create_token():
    auth = new_auth()
    return requests.post(BOOKER_URI + '/auth', json=auth)
# helper function ; refactor the response for update task

def get_booking_ids():
    return requests.get(BOOKER_URI + '/booking/')

def get_booking(booking_id):
    return requests.get(BOOKER_URI + f'/booking/{booking_id}')

def create_booking():
    data = {
        "firstname": "Jacques",
        "lastname": "Revolt",
        "totalprice": 1200,
        "depositpaid": True,
        "bookingdates": {
            "checkin": "2023-01-01",
            "checkout": "2023-01-10"
        },
        "additionalneeds": "test booking"
    }
    return requests.post(BOOKER_URI + '/booking', json=data)

def update_booking(a_token, booking_id):
    headers = {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Cookie': f'token={a_token}'
    }
    data = {
        "firstname": "Jacob",
        "lastname": "Redford",
        "totalprice": 1350,
        "depositpaid": True,
        "bookingdates": {
            "checkin": "2024-01-15",
            "checkout": "2024-01-17"
        },
        "additionalneeds": "Breakfast"
    }
    return requests.put(BOOKER_URI + f'/booking/{booking_id}', headers=headers, json=data)

def partial_update_booking(a_token, booking_id):
    headers = {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Cookie': f'token={a_token}'
    }
    data = {
        "firstname": "James",
        "lastname": "Brown",
    }
    return requests.patch(BOOKER_URI + f'/booking/{booking_id}', headers=headers, json=data)

def delete_booking(a_token, booking_id):
    headers = {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Cookie': f'token={a_token}'
    }
    return requests.delete(BOOKER_URI + f'/booking/{booking_id}', headers=headers)