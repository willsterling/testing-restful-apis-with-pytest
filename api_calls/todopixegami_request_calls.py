from uuid import uuid4

import requests

from config import BASE_URI


def get_task(task_id):
    return requests.get(BASE_URI + f'/get-task/{task_id}')


#refactor the response for create task
def create_task(payload):
    return requests.put(BASE_URI + '/create-task', json=payload)


#refactor the response for update task
def update_task(payload):
    return requests.put(BASE_URI + '/update-task', json=payload)


def list_tasks(user_id):
    return requests.get(BASE_URI + f'/list-tasks/{user_id}')


def delete_task(task_id):
    return requests.delete(BASE_URI + f'/delete-task/{task_id}')


# helper function to create payload
def new_payload():
    # create unique identifier for user id
    # convert to string hex representation of the object returned from uuid4() function
    user_id = f"test_user_{uuid4().hex}"
    task_content = f"test_task_content_{uuid4().hex}"
    return {
        "content": task_content,
        "user_id": user_id,
        "is_done": False
    }
