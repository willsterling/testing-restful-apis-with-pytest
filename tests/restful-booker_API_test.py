import requests
from approvaltests.approvals import verify
from api_calls.restfulbooker_request_calls import new_auth, create_token,get_booking_ids,get_booking,create_booking,update_booking,partial_update_booking, delete_booking


from config import BOOKER_URI


# SECTION: Tests

def get_booking_id():
    response_for_create_booking = create_booking()
    data = response_for_create_booking.json()
    return data["bookingid"]

# HealthCheck
def test_healthcheck_resp_booker():
    # healthcheck endpoint to confirm API is up and running
    response = requests.get(BOOKER_URI + '/ping')
    assert response.status_code == 201

# GetBooking
def test_get_booking():
    booking_id = get_booking_id()

    # get booking from booking id
    response_for_get_booking = get_booking(booking_id)
    assert response_for_get_booking.status_code == 200
    booking = response_for_get_booking.json()

    #assertions for data validation:
    assert booking['firstname'] == 'Jacques'
    assert booking['lastname'] == 'Revolt'
    assert booking['totalprice'] == 1200


# CreateToken
def test_can_create_token():
    # send post request of username and password to create a random token. Assert response is 200
    # auth = new_auth()
    response_for_create_token = create_token()
    assert response_for_create_token.status_code == 200
    data = response_for_create_token.json()

    # verify token is created and is the standard length of 15 characters
    a_token = data['token']
    assert len(a_token) == 15


# GetBookingIds
def test_get_all_bookingids():
    response_for_get_booking_ids = get_booking_ids()
    assert response_for_get_booking_ids.status_code == 200
    booking_ids = response_for_get_booking_ids.json()

    # To see the response data, uncomment:
    # print(booking_ids)
    # print("The current number of bookings is " + str(len(booking_ids)))
    # print("Datatype before deserialization : "+ str(type(booking_ids)))


# CreateBooking
def test_can_create_booking():
    response_for_create_booking = create_booking()
    # print(response_for_create_booking.json())
    assert response_for_create_booking.status_code == 200
    data = response_for_create_booking.json()
    # print(data)


def test_can_update_booking():
    # create auth token, assign token to a variable
    response_for_create_token = create_token()
    a_token = response_for_create_token.json()['token']

    # create a booking, assign booking id to a variable
    booking_id = get_booking_id()

    # update the created booking using token and booking id
    response_for_update_booking = update_booking(a_token, booking_id)
    assert response_for_update_booking.status_code == 200
    booking = response_for_update_booking.json()

    # verification test: returned updated booking information matches sent data in file restful-booker_API_tests.test_can_update_booking.approved.txt
    result = booking
    verify(result)


def test_can_partial_update_booking():
    # create auth token, assign token to a variable
    response_for_create_token = create_token()
    a_token = response_for_create_token.json()['token']

    # create a booking, assign booking id to a variable
    booking_id = get_booking_id()

    response_for_partial_update_booking = partial_update_booking(a_token, booking_id)
    assert response_for_partial_update_booking.status_code == 200


def test_can_delete_booking():
    # create auth token, assign token to a variable
    response_for_create_token = create_token()
    a_token = response_for_create_token.json()['token']

    # create a booking, assign booking id to variable
    booking_id = get_booking_id()

    # delete the booking
    response_for_delete_booking = delete_booking(a_token, booking_id)
    assert response_for_delete_booking.status_code == 201
    # if response_for_delete_booking.status_code == 201:
    #     print("Booking successfully deleted")
    # else:
    #     print("Deletion failed")

