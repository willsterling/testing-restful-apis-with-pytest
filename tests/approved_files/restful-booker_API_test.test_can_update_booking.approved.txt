{'firstname': 'Jacob', 'lastname': 'Redford', 'totalprice': 1350, 'depositpaid': True, 'bookingdates': {'checkin': '2024-01-15', 'checkout': '2024-01-17'}, 'additionalneeds': 'Breakfast'}
